File: Greedy/codeforces_2106D.py

# Problem link ===>> https://codeforces.com/problemset/problem/2106/D

import sys
import math
import bisect
import heapq
from collections import defaultdict, deque, OrderedDict, Counter
from functools import lru_cache, reduce
from itertools import permutations, combinations, accumulate
from typing import List, Tuple, Dict, Set, Optional, Union, Any

MOD = 10**9 + 7
INF = int(1e9 + 5)
input = sys.stdin.read
print = sys.stdout.write

def main():
    data = input().split()
    index = 0
    t = int(data[index])
    index += 1
    
    for _ in range(t):
        n = int(data[index])
        m = int(data[index +1])
        index +=2
        
        a = list(map(int, data[index:index + n]))
        index += n
        b = list(map(int, data[index:index + m]))
        index += m
        
        backward_match = [0] * m
        j = n - 1
        for i in range(m - 1, -1, -1):
            while j >= 0 and a[j] < b[i]:
                j -= 1
            backward_match[i] = j
            j -= 1

        forward_match = [0] * m
        j = 0
        for i in range(m):
            while j < n and a[j] < b[i]:
                j += 1
            forward_match[i] = j
            j += 1

        if forward_match[-1] < n:
            print(0)
            continue

        ans = INF
        for i in range(m):
            match_previous = -1 if i == 0 else forward_match[i - 1]
            match_next = n if i + 1 == m else backward_match[i + 1]
            if match_next > match_previous:
                ans = min(ans, b[i])

        print(-1 if ans == INF else ans)

if __name__ == "__main__":
    main()

File: Greedy/codeforces_2106D.cpp

// Problem link ===>> https://codeforces.com/problemset/problem/2106/D
// submission link ===>> https://codeforces.com/contest/2106/submission/321385918

#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;
#define ll long long
#define ld long double
#define endl '\n'
const int MOD = 1e9 + 7;
const double EPS = 1e-9;
const double PI = acos(-1.0);
const int INF = 1e9 + 5;
template <typename T>
using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);

    ll t;
    cin >> t;
    while (t--)
    {
        ll n, m;
        cin >> n >> m;

        vector<ll> a(n);
        for (ll i = 0; i < n; i++)
            cin >> a[i];

        vector<ll> b(m);
        for (ll i = 0; i < m; i++)
            cin >> b[i];

        vector<ll> backward_match(m);
        ll j = n - 1;
        for (ll i = m - 1; i >= 0; i--)
        {
            while (j >= 0 && a[j] < b[i])
                j--;
            backward_match[i] = j--;
        }

        vector<ll> forward_match(m);
        j = 0;
        for (ll i = 0; i < m; i++)
        {
            while (j < n && a[j] < b[i])
                j++;
            forward_match[i] = j++;
        }

        if (forward_match.back() < n)
        {
            cout << 0 << endl;
            continue;
        }

        ll ans = INF;
        for (ll i = 0; i < m; i++)
        {
            ll match_previous = (i == 0 ? -1 : forward_match[i - 1]);
            ll match_next = (i + 1 == m ? n : backward_match[i + 1]);
            if (match_next > match_previous)
            {
                ans = min(ans, b[i]);
            }
        }

        cout << (ans == INF ? -1 : ans) << endl;
    }

    return 0;
}